hedera-node/hedera-app/src/test/java/com/hedera/node/app/util/HashTreeManagerTest.java

```diff
@@ -0,0 +1,168 @@
+import com.hedera.hapi.block.stream.output.StateChanges;
+import com.hedera.hapi.block.stream.output.TransactionOutput;
+import com.hedera.hapi.block.stream.output.TransactionResult;
+import com.hedera.hapi.node.base.Transaction;
+import com.hedera.pbj.runtime.Codec;
+import com.hedera.pbj.runtime.ParseException;
+import com.hedera.pbj.runtime.io.ReadableSequentialData;
+import com.hedera.pbj.runtime.io.WritableSequentialData;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.common.merkle.MerkleNode;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+
+public class HashTreeManagerTest {
+
+    private final BlockItem blockItem = BlockItem.newBuilder().build();
+    private final Block block = Block.newBuilder().items().build();
+    private final Map<BlockHeader, MerkleNode> inputRootsByBlock = new HashMap<>();
+    private final Map<BlockHeader, MerkleNode> outputRootsByBlock = new HashMap<>();
+    private final String SHA_384_ROOT_HASH =
+            "1333c0960fbb9cac040834be1597b45571dae1bc14b3f78281c20dd147826f017ce4f4105d71ecf394bb41db424a637a";
+
+    @Test
+    public void testHashTreeManager() throws NoSuchAlgorithmException {
+        HashTreeManager<String> tree = new HashTreeManager<>(new SimpleStringCodec());
+        List<String> elements = Arrays.asList("Event1", "Transaction1", "Transaction2", "Event2", "Transaction3");
+        tree.addElements(elements);
+
+        // Calculate the root hash
+        Bytes rootHash = tree.getTreeRoot();
+        assertThat(rootHash).isNotNull();
+        assertThat(rootHash.toString()).isEqualTo(SHA_384_ROOT_HASH);
+    }
+
+    @Test
+    public void testIncrementalConstruction() throws Exception {
+        HashTreeManager<String> tree = new HashTreeManager<>(new SimpleStringCodec());
+
+        tree.addElement("leaf1");
+        // A Merkle tree with a single leaf does have a root hash.
+        assertThat(tree.getTreeRoot()).isNotNull();
+
+        tree.addElement("leaf2");
+        String rootHash = tree.getTreeRootAsString();
+        assertThat(rootHash)
+                .isEqualTo(tree.getTreeRootAsString()); // Root hash should remain consistent after balancing
+
+        tree.addElement("leaf3");
+        assertThat(tree.getTreeRoot()).isNotNull(); // A tree with three leaves still computes a root hash
+
+        tree.addElement("leaf4");
+        String newRootHash = tree.getTreeRootAsString();
+        // Root hash should change as new leaves are added
+        assertThat(rootHash).as("Checking that hashes are not equal").isNotEqualTo(newRootHash);
+    }
+
+    @Test
+    public void testHashBlockItem() throws NoSuchAlgorithmException {
```

Comment:
It may be useful to test the block item-related stuff in a separate class. Or maybe not. Just an idea, not a requirement

Reply:
Yep we can definitely look at that in a follow up PR